File: Sagar_Dataharvesting/Real-Time DB/data harvesting node/marketsocketstream.js
const socketio = require('socket.io-client');
const { createClient } = require('redis');
const fs = require('fs');

class MDSocket_io {
    constructor(token, userID, port, {
        reconnection = true,
        reconnectionAttempts = 0,
        reconnectionDelay = 1,
        reconnectionDelayMax = 50000,
        randomizationFactor = 0.5,
        logger = false,
        binary = false,
        json = null
    } = {}) {
        // Socket.IO client initialization
        this.sid = socketio;
        this.eventListener = this.sid;
        this.marketSocketData = [];
        this.receivedMessageCount = 0;
        
        // Connection parameters
        this.reconnectionDelay = reconnectionDelay;
        this.randomizationFactor = randomizationFactor;
        this.reconnectionDelayMax = reconnectionDelayMax;
        this.port = port;
        this.userID = userID;
        this.token = token;
        this.publishFormat = 'JSON';
        this.broadcastMode = 'Full';
        
        // Connection URL and headers
        this.headers = {
            'Content-Type': 'application/json',
            'Authorization': token
        };
        this.connectionUrl = `${this.port}/?token=${this.token}&userID=${this.userID}&publishFormat=${this.publishFormat}&broadcastMode=${this.broadcastMode}`;
        
        // Redis initialization
        this.redisClient = createClient({
            host: 'localhost',
            port: 6379,
            db: 0
        });
        this.redisMessageCountKey = 'market_data_count';
        
        // Initialize Redis connection
        this.initRedis();
    }

    async initRedis() {
        try {
            await this.redisClient.connect();
            console.log('Connected to Redis');
        } catch (err) {
            console.error('Failed to connect to Redis:', err);
        }
    }

    connect() {
        // Start connection in a non-blocking way
        setTimeout(() => this.startSocketConnection(), 0);
    }

    // startSocketConnection() {
    //     const attemptConnection = () => {
    //         if (!this.socket || !this.socket.connected) {
    //             try {
    //                 console.log('Attempting to connect to market data socket...');
    //                 this.socket = this.sid.connect(this.connectionUrl, {
    //                     headers: this.headers,
    //                     transports: ['websocket'],
    //                     path: '/apimarketdata/socket.io',
    //                     reconnection: true,
    //                     reconnectionAttempts: Infinity,
    //                     reconnectionDelay: this.reconnectionDelay,
    //                     reconnectionDelayMax: this.reconnectionDelayMax,
    //                     randomizationFactor: this.randomizationFactor
    //                 });

    //                 this.registerEventHandlers();
    //             } catch (error) {
    //                 console.error('Connection error:', error);
    //                 setTimeout(attemptConnection, 5000); // Retry after 5 seconds
    //             }
    //         }
    //     };

    //     attemptConnection();
    // }
    startSocketConnection() {
        const attemptConnection = () => {
            if (!this.socket || !this.socket.connected) {
                try {
                    console.log('Detailed Connection Parameters:');
                    console.log('Full URL:', this.connectionUrl);
                    console.log('Token:', this.token);
                    console.log('User ID:', this.userID);
                    console.log('Publish Format:', this.publishFormat);
                    console.log('Broadcast Mode:', this.broadcastMode);
    
                    // More comprehensive connection configuration
                    this.socket = this.sid.connect(this.port, {
                        reconnection: true,
                        reconnectionAttempts: Infinity,
                        reconnectionDelay: this.reconnectionDelay,
                        reconnectionDelayMax: this.reconnectionDelayMax,
                        randomizationFactor: this.randomizationFactor,
                        transports: ['websocket'],
                        timeout: 10000,
                        forceNew: true,
                        path: '/apimarketdata/socket.io', // Specific path if required
                        query: {
                            token: this.token,
                            userID: this.userID,
                            publishFormat: this.publishFormat,
                            broadcastMode: this.broadcastMode
                        },
                        extraHeaders: {
                            'Authorization': this.token,
                            'Content-Type': 'application/json'
                        }
                    });
    
                    // Comprehensive event logging
                    this.socket.on('connect', () => {
                        console.log('Socket Connected Successfully');
                        console.log('Socket ID:', this.socket.id);
                    });
    
                    this.socket.on('disconnect', (reason) => {
                        console.log('Socket Disconnected. Reason:', reason);
                    });
    
                    this.socket.on('connect_error', (error) => {
                        console.error('Connection Error Details:', {
                            name: error.name,
                            message: error.message,
                            stack: error.stack
                        });
                    });
    
                    this.socket.on('reconnect_attempt', (attemptNumber) => {
                        console.log(`Attempting to reconnect. Attempt: ${attemptNumber}`);
                    });
    
                    this.registerEventHandlers();
                } catch (error) {
                    console.error('Socket Connection Initialization Catastrophic Error:', error);
                    setTimeout(attemptConnection, 5000);
                }
            }
        };
    
        attemptConnection();
    }
    registerEventHandlers() {
        if (!this.socket) return;

        this.socket.on('connect', () => {
            console.log('Market Data Socket connected successfully!');
            console.log(`socket reconnected @ ${new Date().toISOString()}`);
        });

        this.socket.on('1501-json-full', (data) => {
            console.log('Received 1501 Level1, Touchline message!', data);
        });

        this.socket.on('1512-json-full', async (data) => {
            // console.log('Received 1512 Level2, Touchline message!', data);
            this.receivedMessageCount += 1;
            const jsonData = typeof data === 'string' ? JSON.parse(data) : data;
            try {
                await this.redisClient.rPush('market_data', JSON.stringify(jsonData));
                await this.redisClient.set(this.redisMessageCountKey, this.receivedMessageCount);
                
                // Log the receivedMessageCount to datacounter.txt
                fs.writeFile('datacounter.txt', `Received Message Count: ${this.receivedMessageCount}`, (err) => {
                    if (err) {
                        console.error('Error writing to datacounter.txt:', err);
                    }
                });
            } catch (err) {
                console.error('Error saving to Redis:', err);
            }
        });

        this.socket.on('disconnect', () => {
            console.log('Market Data Socket disconnected!');
            console.log(`socket disconnected @ ${new Date().toISOString()}`);
        });

        this.socket.on('error', (data) => {
            console.log('Market Data Error:', data);
        });
    }

    async verifyMessageCounts() {
        try {
            const redisMessageCount = parseInt(await this.redisClient.get(this.redisMessageCountKey)) || 0;
            if (this.receivedMessageCount === redisMessageCount) {
                console.log(`All messages are successfully saved to Redis! Count: ${this.receivedMessageCount}`);
            } else {
                console.log(`Data mismatch! Received: ${this.receivedMessageCount}, Saved in Redis: ${redisMessageCount}`);
            }
        } catch (err) {
            console.error('Error verifying message counts:', err);
        }
    }

    getEmitter() {
        return this.eventListener;
    }

    disconnect() {
        if (this.socket) {
            this.socket.disconnect();
        }
        this.redisClient.quit();
    }
}

module.exports = MDSocket_io;

File: Sagar_Dataharvesting/Real-Time DB/data_harvesting 2.0/MarketSocketStream.py
import threading
import socketio
import json
from datetime import datetime
from utils import Logger
import time
import redis

logger = Logger('socket_log.txt')

class MDSocket_io:
    def __init__(self, token, userID, port, publisher, reconnection=True, reconnection_attempts=0, reconnection_delay=1,
                 reconnection_delay_max=50000, randomization_factor=0.5, logger=False, binary=False, json=None,
                 **kwargs):
        self.publisher = publisher
        self.sid = socketio.Client(logger=False, engineio_logger=False)
        self.eventlistener = self.sid
        self.market_socket_data = []
        self.received_message_count = 0
        self.register_event_handlers()
        self.reconnection_delay = reconnection_delay
        self.randomization_factor = randomization_factor
        self.reconnection_delay_max = reconnection_delay_max
        self.port = port
        self.userID = userID
        self.token = token
        self.publishFormat = 'JSON'
        self.broadcastMode = "Full"
        self.headers = {'Content-Type': 'application/json', 'Authorization': token}
        self.connection_url = f'{self.port}/?token={self.token}&userID={self.userID}&publishFormat={self.publishFormat}&broadcastMode={self.broadcastMode}'
        self.redis_client = redis.Redis(host='localhost', port=6379, db=0)
        self.redis_message_count_key = 'market_data_count'
        self.namespaces_lock = threading.Lock()
    def connect(self):
        threading.Thread(target=self.start_socket_connection).start()

    def start_socket_connection(self):
        while True:
            if not self.sid.connected:
                try:
                    try:
                        self.sid.connect(
                            url=self.connection_url,
                            headers=self.headers,
                            transports='websocket',
                            namespaces=None,
                            socketio_path='/apimarketdata/socket.io'
                        )
                        self.sid.wait()
                    except:
                        print("already connected")
                except socketio.exceptions.ConnectionError as e:
                    pass
            else:
                pass

    def register_event_handlers(self):
        self.sid.on('connect', self.on_connect)
        self.sid.on('disconnect', self.on_disconnect)
        self.sid.on('error', self.on_error)
        self.sid.on('1501-json-full', self.on_message1501_json_full)
        self.sid.on('1512-json-full', self.on_message1512_json_full)
        self.sid.on('disconnect', self.on_disconnect)
        self.sid.on('error', self.on_error)

    def on_connect(self):
        print('Market Data Socket connected successfully!')
        logger.log(f'socket reconnected @ {datetime.now()}')

    def on_message1501_json_full(self, data):
        print('I received a 1501 Level1, Touchline message!', data)

    def on_message1512_json_full(self, data):
        self.received_message_count += 1
        data = json.loads(data)
        # print(data)
        self.redis_client.rpush('market_data', json.dumps(data))
        self.redis_client.incr(self.redis_message_count_key)

    def verify_message_counts(self):
        redis_message_count = int(self.redis_client.get(self.redis_message_count_key) or 0)
        if self.received_message_count == redis_message_count:
            print(f"All messages are successfully saved to Redis! Count: {self.received_message_count}")
        else:
            print(f"Data mismatch! Received: {self.received_message_count}, Saved in Redis: {redis_message_count}")

    def on_disconnect(self):
        with self.namespaces_lock:
            try:
                print('Market Data Socket disconnected!')
                logger.log(f'socket disconnected @ {datetime.now()}')
            except RuntimeError as e:
                print(f"Disconnect handler error: {e}")

    def on_error(self, data):
        print('Market Data Error:', data)

    def get_emitter(self):
        return self.eventlistener

    def disconnect(self):
        self.sid.disconnect()
